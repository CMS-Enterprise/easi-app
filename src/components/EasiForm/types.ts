import React from 'react';
import {
  FieldPath,
  FieldValues,
  RefCallBack,
  RegisterOptions,
  UseFormRegisterReturn,
  UseFormReturn
} from 'react-hook-form';

export type EasiPartialSubmitProps<TFieldValues extends FieldValues> = {
  update: (
    /** Object containing valid dirty field values */
    formData: Partial<TFieldValues>
  ) => Promise<any>;
  /** Callback to execute after updating */
  callback?: () => void;
  /** Whether to clear field error messages */
  clearErrors?: boolean;
};

export type EasiPartialSubmit<TFieldValues extends FieldValues> = ({
  update,
  clearErrors,
  callback
}: EasiPartialSubmitProps<TFieldValues>) => Promise<void>;

export type EasiFormRegister<TFieldValues extends FieldValues> = <
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
>(
  name: TFieldName,
  options?: RegisterOptions<TFieldValues, TFieldName>
) => UseFormRegisterReturn<TFieldName> & { inputRef: RefCallBack };

export type UseEasiFormReturn<
  TFieldValues extends FieldValues = FieldValues,
  TContext = any
> = Omit<UseFormReturn<TFieldValues, TContext>, 'register'> & {
  /** Ignore errors and submit valid dirty field values */
  partialSubmit: EasiPartialSubmit<TFieldValues>;
  /** Custom register function that returns `inputRef` for function components */
  register: EasiFormRegister<TFieldValues>;
};

export type EasiFormProviderProps<
  TFieldValues extends FieldValues,
  TContext = any
> = {
  children: React.ReactNode | React.ReactNode[];
} & UseEasiFormReturn<TFieldValues, TContext>;
