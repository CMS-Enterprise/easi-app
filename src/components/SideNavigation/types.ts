export type SubNavLinkProps = {
  route: string;
  text: string;
};

export type NavLinkProps = SubNavLinkProps & {
  children?: SubNavLinkProps[];
  /** Designates end of navigation group with bottom border */
  groupEnd?: boolean;
};
