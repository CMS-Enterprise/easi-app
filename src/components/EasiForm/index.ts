export { default as useEasiForm } from './useEasiForm';
export { default as useEasiFormContext } from './useEasiFormContext';
export { default as EasiFormProvider } from './EasiFormProvider';
