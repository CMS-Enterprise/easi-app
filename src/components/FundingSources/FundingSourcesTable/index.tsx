import React, { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { Column, useTable } from 'react-table';
import { Button, Table } from '@trussworks/react-uswds';

import { FormattedFundingSource } from 'types/systemIntake';

type FundingSourcesTableProps = {
  fundingSources: FormattedFundingSource[];
  removeFundingSource: (index: number) => void;
};

const FundingSourcesTable = ({
  fundingSources,
  removeFundingSource
}: FundingSourcesTableProps) => {
  const { t } = useTranslation('fundingSources');

  const columns = useMemo<Column<FormattedFundingSource>[]>(() => {
    return [
      {
        Header: t<string>('projectNumber'),
        accessor: 'projectNumber',
        id: 'projectNumber'
      },
      {
        Header: t<string>('investments'),
        accessor: 'investments',
        id: 'investments',
        Cell: ({ value }: { value: string[] }) => <>{value.join(', ')}</>
      },
      {
        Header: t<string>('general:actions'),
        id: 'actions',
        accessor: (_, index) => {
          return (
            <Button
              unstyled
              type="button"
              className="text-error margin-y-0"
              onClick={() => removeFundingSource(index)}
              data-testid={`removeFundingSource-${index}`}
            >
              {t('general:remove')}
            </Button>
          );
        }
      }
    ];
  }, [t, removeFundingSource]);

  const table = useTable({
    columns,
    data: fundingSources,
    autoResetSortBy: false,
    autoResetPage: true
  });

  const { getTableBodyProps, getTableProps, headerGroups, prepareRow, rows } =
    table;

  return (
    <div>
      <Table bordered={false} fullWidth scrollable {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => {
            const { key, ...headerGroupProps } =
              headerGroup.getHeaderGroupProps();

            return (
              <tr {...headerGroupProps} key={key}>
                {headerGroup.headers.map(column => {
                  const { key: headerKey, ...headerProps } =
                    column.getHeaderProps();

                  return (
                    <th {...headerProps} key={headerKey}>
                      {column.render('Header')}
                    </th>
                  );
                })}
              </tr>
            );
          })}
        </thead>

        <tbody {...getTableBodyProps()}>
          {rows.map(row => {
            prepareRow(row);
            const { key: rowKey, ...rowProps } = row.getRowProps();

            return (
              <tr {...rowProps} key={rowKey}>
                {row.cells.map(cell => {
                  const { key: cellKey, ...cellProps } = cell.getCellProps();

                  return (
                    <td {...cellProps} key={cellKey}>
                      {cell.render('Cell')}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </Table>
    </div>
  );
};

export default FundingSourcesTable;
