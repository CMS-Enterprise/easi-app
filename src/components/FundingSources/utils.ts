import { FundingSource } from 'types/systemIntake';

import { FormattedFundingSource } from '.';

/** Formats funding sources for API */
export const formatFundingSourcesForApi = (
  fundingSources: FormattedFundingSource[]
): FundingSource[] => {
  return fundingSources
    .map(({ id, fundingNumber, sources }) => {
      return sources.map(source => ({
        id,
        fundingNumber,
        source
      }));
    })
    .flat();
};

/** Formats funding sources for app by grouping objects by funding number */
export const formatFundingSourcesForApp = (
  fundingSources: FundingSource[]
): FormattedFundingSource[] => {
  return fundingSources.reduce<FormattedFundingSource[]>(
    (acc, { id, fundingNumber, source }) => {
      const existingSource = acc.find(s => s.fundingNumber === fundingNumber);

      if (source && existingSource) {
        existingSource.sources.push(source);
        return acc;
      }

      return [
        ...acc,
        {
          id,
          fundingNumber,
          sources: source ? [source] : []
        }
      ];
    },
    []
  );
};
