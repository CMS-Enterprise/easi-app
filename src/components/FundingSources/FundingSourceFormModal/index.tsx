import React, { useEffect, useMemo } from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { ErrorMessage } from '@hookform/error-message';
import { yupResolver } from '@hookform/resolvers/yup';
import {
  Button,
  ButtonGroup,
  Fieldset,
  Label,
  ModalFooter,
  ModalHeading,
  TextInput
} from '@trussworks/react-uswds';

import { useEasiForm } from 'components/EasiForm';
import ExternalLinkAndModal from 'components/ExternalLinkAndModal';
import FieldErrorMsg from 'components/FieldErrorMsg';
import FieldGroup from 'components/FieldGroup';
import HelpText from 'components/HelpText';
import Modal from 'components/Modal';
import MultiSelect from 'components/MultiSelect';
import RequiredFieldsText from 'components/RequiredFieldsText';
import intakeFundingSources from 'constants/enums/intakeFundingSources';
import { FormattedFundingSource } from 'types/systemIntake';
import FundingSourceValidationSchema from 'validations/fundingSources';

type FundingSourcesModalProps = {
  isOpen: boolean;
  closeModal: () => void;
  addFundingSource: (data: FormattedFundingSource) => void;
  initialFundingSources: FormattedFundingSource[];
};

/**
 * Modal that contains the form for adding a funding source to a system intake
 */
const FundingSourceFormModal = ({
  isOpen,
  closeModal,
  addFundingSource,
  initialFundingSources
}: FundingSourcesModalProps) => {
  const { t } = useTranslation('fundingSources');

  const {
    control,
    handleSubmit,
    register,
    reset,
    watch,
    formState: { errors }
  } = useEasiForm<FormattedFundingSource>({
    defaultValues: {
      projectNumber: '',
      investments: []
    },
    reValidateMode: 'onSubmit',
    resolver: yupResolver(FundingSourceValidationSchema),
    // Pass initial project numbers as context to resolver to validate uniqueness
    context: {
      initialProjectNumbers: initialFundingSources.map(
        source => source.projectNumber
      )
    }
  });

  const submitForm = handleSubmit(data => {
    addFundingSource(data);
    closeModal();
  });

  const projectNumber = watch('projectNumber');
  const investments = watch('investments');

  /** Disable form submit if required fields are blank */
  const disableSubmit = useMemo(() => {
    return !projectNumber || investments.length === 0;
  }, [projectNumber, investments]);

  // Reset form on modal close
  useEffect(() => {
    if (!isOpen) {
      reset();
    }
  }, [isOpen, reset]);

  return (
    <Modal isOpen={isOpen} closeModal={closeModal} className="font-body-md">
      <Fieldset>
        <legend>
          <ModalHeading className="margin-bottom-0">
            {t('addFundingSource')}
          </ModalHeading>
        </legend>

        <RequiredFieldsText className="margin-y-0" />

        <FieldGroup className="margin-top-2" error={!!errors?.projectNumber}>
          <Label htmlFor="projectNumber" className="text-normal" requiredMarker>
            {t('projectNumber')}
          </Label>
          <HelpText id="projectNumberHelpText">
            {t('form.projectNumberHelpText')}
          </HelpText>
          <ErrorMessage
            errors={errors}
            name="projectNumber"
            as={FieldErrorMsg}
          />
          <TextInput
            {...register(`projectNumber`)}
            type="text"
            id="projectNumber"
            data-testid="projectNumberInput"
            className="maxw-none margin-bottom-1"
            aria-describedby="projectNumberHelptext projectNumberHelpLink"
          />

          <ExternalLinkAndModal href="https://cmsintranet.share.cms.gov/JT/Pages/Budget.aspx">
            <span id="projectNumberHelpLink">
              {t('form.projectNumberLink')}
            </span>
          </ExternalLinkAndModal>
        </FieldGroup>

        <FieldGroup error={!!errors?.investments}>
          <Label htmlFor="investments" className="text-normal" requiredMarker>
            {t('investment')}
          </Label>
          <ErrorMessage errors={errors} name="investments" as={FieldErrorMsg} />
          <Controller
            name="investments"
            control={control}
            render={({ field: { ref, ...field } }) => (
              <MultiSelect
                {...field}
                id="investments"
                data-testid="investmentsInput"
                selectedLabel={t('form.selectedInvestments')}
                options={intakeFundingSources.map(option => ({
                  value: option,
                  label: t(option)
                }))}
              />
            )}
          />
        </FieldGroup>

        <ModalFooter>
          <ButtonGroup>
            <Button
              type="button"
              onClick={submitForm}
              className="margin-right-2"
              disabled={disableSubmit}
              data-testid="submitFundingSourceButton"
            >
              {t('form.addFundingSource')}
            </Button>
            <Button
              type="button"
              onClick={closeModal}
              unstyled
              className="text-error"
            >
              {t('Cancel')}
            </Button>
          </ButtonGroup>
        </ModalFooter>
      </Fieldset>
    </Modal>
  );
};

export default FundingSourceFormModal;
