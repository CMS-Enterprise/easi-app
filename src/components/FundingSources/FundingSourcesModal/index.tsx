import React, { useEffect } from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { ErrorMessage } from '@hookform/error-message';
import { yupResolver } from '@hookform/resolvers/yup';
import {
  Button,
  ButtonGroup,
  Fieldset,
  Label,
  Link,
  ModalFooter,
  ModalHeading,
  TextInput
} from '@trussworks/react-uswds';

import { useEasiForm } from 'components/EasiForm';
import FieldErrorMsg from 'components/FieldErrorMsg';
import FieldGroup from 'components/FieldGroup';
import HelpText from 'components/HelpText';
import Modal from 'components/Modal';
import MultiSelect from 'components/MultiSelect';
import RequiredFieldsText from 'components/RequiredFieldsText';
import intakeFundingSources from 'constants/enums/intakeFundingSources';
import { FormattedFundingSource } from 'types/systemIntake';
import { FundingSourceValidationSchema } from 'validations/systemIntakeSchema';

type FundingSourcesModalProps = {
  isOpen: boolean;
  closeModal: () => void;
  addFundingSource: (data: FormattedFundingSource) => void;
};

const FundingSourcesModal = ({
  isOpen,
  closeModal,
  addFundingSource
}: FundingSourcesModalProps) => {
  const { t } = useTranslation('intake');

  const {
    control,
    handleSubmit,
    register,
    reset,
    formState: { errors }
  } = useEasiForm<FormattedFundingSource>({
    resolver: yupResolver(FundingSourceValidationSchema),
    defaultValues: {
      projectNumber: '',
      investments: []
    }
  });

  const submitForm = handleSubmit(data => {
    addFundingSource(data);
    closeModal();
  });

  // Reset form on modal close
  useEffect(() => {
    if (!isOpen) {
      reset();
    }
  }, [isOpen, reset]);

  return (
    <Modal isOpen={isOpen} closeModal={closeModal} className="font-body-md">
      <Fieldset>
        <legend>
          <ModalHeading className="margin-bottom-0">
            {t('contractDetails.fundingSources.addFundingSource')}
          </ModalHeading>
        </legend>

        <RequiredFieldsText className="margin-y-0" />

        <FieldGroup className="margin-top-2" error={!!errors?.projectNumber}>
          <Label htmlFor="projectNumber" className="text-normal" requiredMarker>
            {t('contractDetails.fundingSources.projectNumber')}
          </Label>
          <HelpText id="projectNumberHelpText">
            {t('contractDetails.fundingSources.projectNumberHelpText')}
          </HelpText>
          <ErrorMessage
            errors={errors}
            name="projectNumber"
            as={FieldErrorMsg}
          />
          <TextInput
            {...register(`projectNumber`)}
            type="text"
            id="projectNumber"
            className="maxw-none"
            aria-describedby="projectNumberHelptext projectNumberHelpLink"
          />

          <HelpText id="projectNumberHelpLink" className="margin-top-1">
            <Link
              href="https://cmsintranet.share.cms.gov/JT/Pages/Budget.aspx"
              target="_blank"
              rel="noopener noreferrer"
              variant="external"
            >
              {t('contractDetails.fundingSources.projectNumberLink')}
            </Link>
          </HelpText>
        </FieldGroup>

        <FieldGroup error={!!errors?.investments}>
          <Label htmlFor="investments" className="text-normal" requiredMarker>
            {t('contractDetails.fundingSources.investment')}
          </Label>
          <ErrorMessage errors={errors} name="investments" as={FieldErrorMsg} />
          <Controller
            name="investments"
            control={control}
            render={({ field: { ref, ...field } }) => (
              <MultiSelect
                {...field}
                id="investments"
                selectedLabel={t(
                  'contractDetails.fundingSources.selectedInvestments'
                )}
                options={intakeFundingSources.map(option => ({
                  value: option,
                  label: t(option)
                }))}
              />
            )}
          />
        </FieldGroup>

        <ModalFooter>
          <ButtonGroup>
            <Button
              type="button"
              onClick={submitForm}
              className="margin-right-2"
            >
              {t('contractDetails.fundingSources.modalSubmit')}
            </Button>
            <Button
              type="button"
              onClick={closeModal}
              unstyled
              className="text-error"
            >
              {t('Cancel')}
            </Button>
          </ButtonGroup>
        </ModalFooter>
      </Fieldset>
    </Modal>
  );
};

export default FundingSourcesModal;
