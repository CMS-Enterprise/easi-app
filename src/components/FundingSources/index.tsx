import React, { useEffect, useState } from 'react';
import { Controller, useFieldArray } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { Button } from '@trussworks/react-uswds';

import CheckboxField from 'components/CheckboxField';
import { useEasiFormContext } from 'components/EasiForm';
import { ContractDetailsForm } from 'types/systemIntake';

import FundingSourcesModal from './FundingSourcesModal';

/** Funding sources component for system intake form */
const FundingSources = () => {
  const { t } = useTranslation('intake');

  const [isModalOpen, setIsModalOpen] = useState(false);

  const { control, watch } = useEasiFormContext<ContractDetailsForm>();

  const arrayfield = useFieldArray({
    control,
    name: 'fundingSources'
  });

  const { fields, append, remove } = arrayfield;

  const skipFundingSources = watch('skipFundingSources');

  /** Remove all funding sources if checkbox is checked */
  useEffect(() => {
    if (skipFundingSources) {
      remove();
    }
  }, [skipFundingSources, remove]);

  return (
    <div id="intakeFundingSources">
      <FundingSourcesModal
        isOpen={isModalOpen}
        closeModal={() => setIsModalOpen(false)}
        addFundingSource={fundingSource => append(fundingSource)}
        initialFundingSources={fields}
      />

      {fields.map((source, index) => {
        const { projectNumber, investments } = source;

        // Display funding source
        return (
          <div
            id={`fundingSource-${index}`}
            // eslint-disable-next-line react/no-array-index-key
            key={`fundingSource-${index}`}
            className="margin-top-205"
          >
            <p className="text-bold font-body-sm margin-bottom-0">
              {t('contractDetails.fundingSources.investment')}
            </p>
            <p className="margin-y-05">
              {t('contractDetails.fundingSources.projectNumberLabel', {
                projectNumber
              })}
            </p>
            <p className="margin-y-05">
              {t('contractDetails.fundingSources.investmentsLabel', {
                investments: investments.join(', ')
              })}
            </p>

            <Button
              unstyled
              onClick={() => remove(index)}
              type="button"
              className="text-error margin-top-1"
            >
              {t('Delete')}
            </Button>
          </div>
        );
      })}

      <Button
        type="button"
        onClick={() => setIsModalOpen(true)}
        disabled={skipFundingSources}
        className="margin-top-2"
        data-testid="fundingSourcesAction-add"
        outline
      >
        {t(
          `contractDetails.fundingSources.${
            fields.length > 0 ? 'addAnotherFundingSource' : 'addFundingSource'
          }`
        )}
      </Button>

      <Controller
        control={control}
        name="skipFundingSources"
        render={({ field: { ref, ...field } }) => {
          return (
            <CheckboxField
              {...field}
              id={field.name}
              value="true"
              onChange={e => field.onChange(e.target.checked)}
              label={t(
                'contractDetails.fundingSources.skipFundingSourcesLabel'
              )}
            />
          );
        }}
      />
    </div>
  );
};

export default FundingSources;
