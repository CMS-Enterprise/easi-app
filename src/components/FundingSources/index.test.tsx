import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { EasiFormProvider, useEasiForm } from 'components/EasiForm';
import { FormattedFundingSource } from 'types/systemIntake';

import FundingSources from '.';

type FormType = { fundingSources: FormattedFundingSource[] };

describe('Funding sources', () => {
  /** Returns form provider for unit tests */
  const Wrapper = ({
    children,
    fundingSources = []
  }: FormType & {
    children: React.ReactNode;
  }) => {
    const form = useEasiForm<FormType>({
      defaultValues: {
        fundingSources
      }
    });

    return <EasiFormProvider<FormType> {...form}>{children}</EasiFormProvider>;
  };

  it('renders with no funding sources', () => {
    render(
      <Wrapper fundingSources={[]}>
        <FundingSources />
      </Wrapper>
    );

    expect(
      screen.getByRole('button', { name: 'Add a funding source' })
    ).toBeInTheDocument();
  });

  it('renders error messages', async () => {
    render(
      <Wrapper
        fundingSources={[
          {
            projectNumber: '123456',
            investments: ['Fed Admin']
          }
        ]}
      >
        <FundingSources />
      </Wrapper>
    );

    userEvent.click(
      screen.getByRole('button', { name: 'Add another funding source' })
    );

    expect(await screen.findByRole('dialog')).toBeInTheDocument();

    userEvent.click(
      screen.getByRole('button', {
        name: 'Add funding source'
      })
    );

    // Button is disabled with empty fields
    expect(
      screen.getByRole('button', { name: 'Add funding source' })
    ).toBeDisabled();

    // Add investment
    userEvent.type(screen.getByRole('combobox'), 'Fed Admin {enter}');
    expect(await screen.findByText('1 selected')).toBeInTheDocument();

    // Check funding number is numeric

    const projectNumberField = screen.getByRole('textbox', {
      name: 'Project number *'
    });

    userEvent.type(projectNumberField, 'aaaaaa');
    expect(projectNumberField).toHaveValue('aaaaaa');

    userEvent.click(screen.getByRole('button', { name: 'Add funding source' }));

    expect(
      await screen.findByText('Project number can only contain digits')
    ).toBeInTheDocument();

    // Check unique funding number

    userEvent.clear(projectNumberField);
    userEvent.type(projectNumberField, '123456');
    expect(projectNumberField).toHaveValue('123456');

    userEvent.click(screen.getByRole('button', { name: 'Add funding source' }));

    expect(
      await screen.findByText(
        'Project number has already been added to this request'
      )
    ).toBeInTheDocument();
  });
});
