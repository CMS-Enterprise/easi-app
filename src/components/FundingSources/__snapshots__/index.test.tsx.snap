// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Funding sources > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    id="intakeFundingSources"
  >
    <button
      class="usa-button usa-button--outline margin-top-2"
      data-testid="addFundingSourceButton"
      type="button"
    >
      Add another funding source
    </button>
    <div
      class="easi-checkbox usa-checkbox"
    >
      <input
        class="usa-checkbox__input"
        data-testid="clearFundingSourcesCheckbox"
        id="existingFunding"
        name="existingFunding"
        type="checkbox"
        value="true"
      />
      <label
        class="usa-checkbox__label"
        for="existingFunding"
      >
        or, check this box if a new investment and/or project number is still being set up for this project
      </label>
    </div>
    <div
      class="usa-table-container--scrollable"
      data-testid="scrollable-table-container"
      tabindex="0"
    >
      <table
        class="usa-table usa-table--borderless width-full"
        data-testid="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="width: 250px;"
            >
              <button
                class="usa-button usa-button--unstyled width-full flex-justify margin-y-0"
                data-testid="button"
                style="cursor: pointer;"
                title="Toggle SortBy"
                type="button"
              >
                Project number
                <svg
                  aria-hidden="true"
                  class="usa-icon margin-left-05"
                  data-testid="caret--sort"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="width: auto;"
            >
              <button
                class="usa-button usa-button--unstyled width-full flex-justify margin-y-0"
                data-testid="button"
                style="cursor: pointer;"
                title="Toggle SortBy"
                type="button"
              >
                Investment(s)
                <svg
                  aria-hidden="true"
                  class="usa-icon margin-left-05"
                  data-testid="caret--sort"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                  />
                </svg>
              </button>
            </th>
            <th
              colspan="1"
              role="columnheader"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            data-testid="fundingSource123456"
            role="row"
          >
            <td
              role="cell"
            >
              123456
            </td>
            <td
              role="cell"
            >
              Fed Admin
            </td>
            <td
              role="cell"
            >
              <button
                class="usa-button usa-button--unstyled text-error margin-y-0"
                data-testid="removeFundingSource-0"
                type="button"
              >
                Remove
              </button>
            </td>
          </tr>
          <tr
            data-testid="fundingSource789012"
            role="row"
          >
            <td
              role="cell"
            >
              789012
            </td>
            <td
              role="cell"
            >
              HITECH Medicare
            </td>
            <td
              role="cell"
            >
              <button
                class="usa-button usa-button--unstyled text-error margin-y-0"
                data-testid="removeFundingSource-1"
                type="button"
              >
                Remove
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
